### ISSUE TITLE
False positive for DMI_RANDOM_USED_ONLY_ONCE when used in static { } initializer

### ISSUE THREAD

**Issue by sratz** [spotbugs/spotbugs#3931]
**Describe the bug**
`DMI_RANDOM_USED_ONLY_ONCE` produces a false positives when using a `Random` inside of a `static { }` initializer.

**To Reproduce**

```java
import java.security.SecureRandom;
import java.util.Random;

public class Bug {
	static final Random random = new SecureRandom();
	static final byte[] seed = new byte[32];
	static {
		random.nextBytes(seed);
	}
}
```
```
.../Bug.java:8 Random object created and used only once in Bug.<static initializer for Bug>() [Troubling(14), High confidence]
```

**Expected behavior**
No warning.

**Environment**
- Java Compiler: ECJ (Eclipse 2026-03)
- Java version: JDK 21
- SpotBugs Version: 4.9.8.2
- How do you use SpotBugs: Eclipse; Maven Plugin

**Additional context**
If I extract things to a method, the warning disappears, although this is functionally equivalent:
```java
import java.security.SecureRandom;
import java.util.Random;

public class Bug {
	static final Random random = new SecureRandom();
	static final byte[] seed = new byte[32];
	static {
		seed();
	}

	private static void seed() {
		random.nextBytes(seed);
	}
}
```

or also

```java
import java.security.SecureRandom;
import java.util.Random;

public class Bug {
	static final Random random = new SecureRandom();
	static final byte[] seed = seed();

	private static byte[] seed() {
		byte[] seed = new byte[32];
		random.nextBytes(seed);
		return seed;
	}
}
```

**Comment by JuditKnoll** [comment]
To me this looks like a true positive. From the examples it looks like to me that random is only used once. Is there something I'm missing here?

**Comment by sratz** [comment]
> To me this looks like a true positive. From the examples it looks like to me that random is only used once. Is there something I'm missing here?

Yes it is only used once (I only need it once) but even if I use it twice it won't change the warning.

The problem is that the check finds occurrences where the secret is **_created and used in the same method_**:

_.../Bug.java:8 Random object **created and used only once in** Bug.<static initializer for Bug>() [Troubling(14), High confidence]_

And technically the static assignment before the `static {}` block and the block itself are the same thing in the byte code, hence it treats this static initializer as one 'method' in which the Random is both created and used - and only used there. This is why pulling it out into a different method avoids the warning.

I think the static initializer should be handled specially.

**Comment by JuditKnoll** [comment]
This is not a false positive then, but a false negative when the usage is in a separate private method, which is only called once.
Can you please share the code example where it's used more than once, and the error is still reported? Is it like https://github.com/spotbugs/spotbugs/issues/3830?

Can you please elaborate why should the static initializer handled specially?

**Comment by sratz** [comment]
What I am saying is: The fact that the Random is created as a static fields should already make this warning not occur:
- It's the one and only instance of Random in the whole class
- It is only created once
- It should not matter how often and from where it is used

**Comment by JuditKnoll** [comment]
From what I can see, this looks like a true positive and I find it unlikely that the detector will be modified.
I suggest suppressing it by the annotation in the code or filtering it out via filter file.
